[WIP] Support off action

kattz-kawa opened the pull request:
This new draft PR follows up on #124 and was created because a rebase was needed.  
The content and purpose of this PR are the same as #124.

[ECOPROJECT-1471](https://issues.redhat.com/browse/ECOPROJECT-1471)

kattz-kawa commented:
@mshitrit, Thank you for the feedback. That makes a lot of sense.

In this case, I wasn’t able to push to the original branch, which is why I opened this new PR to continue the work on behalf of @k-keiichi-rh.  
I understand that keeping the original context is important, and I’ll keep that in mind for future contributions.

kattz-kawa commented:
I saw the rebase request. I've enabled "Allow edits by maintainers" on this PR, so please feel free to make any necessary changes or rebase directly if that's easier. Let me know if you'd prefer me to handle it instead. Thanks!

mshitrit commented:
@coderabbitai full review

mshitrit commented on controllers/fenceagentsremediation_controller.go:
```diff
@@ -234,6 +236,7 @@ func (r *FenceAgentsRemediationReconciler) Reconcile(ctx context.Context, req ct
 		faParams, isRetryRequired, err := r.buildFenceAgentParams(ctx, far)
 		if err != nil {
 			if !isRetryRequired {
+				commonEvents.WarningEvent(r.Recorder, far, utils.EventReasonCrParamInvalid, utils.EventMessageCrParamInvalid)
```
This might be a bit confusing since there are currently 3 use cases which can cause this:
1. Duplication of a param (which I think this event matches reasonably)
2. Invalid **value** of the action param (which I'm not sure is a great match)
3. missing params (which I think this event does not match at all)

kattz-kawa replied:
Thank you for pointing that out. I’ve updated the code.

mshitrit replied:
Hmm I think I'm missing something, I could see a change.
Let me know if I'm missing something obvious.

1. Invalid value of the action param would still trigger this event, see [here](https://github.com/medik8s/fence-agents-remediation/pull/175/files#diff-05147ee3f0d225b5a60754b8b9bff2c569deb0a9d7eaddae33af680291d493a1R455-R457)
2. Same for missing params. (see [here](https://github.com/medik8s/fence-agents-remediation/pull/175/files#diff-05147ee3f0d225b5a60754b8b9bff2c569deb0a9d7eaddae33af680291d493a1R496-R500))

mshitrit commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +221,134 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+func GetPowerStatus(ctx *FenceCommandContext, targetNodeName string) (string, error) {
+	log := ctx.Logger
+
+	far_controller, err := GetActiveFARControllerManagerPod(ctx.ClientSet, ctx.ControllerNS, ctx.Logger)
```
far_controller does not match camelCase naming convention

kattz-kawa replied:
I’ve renamed far_controller to farController.

mshitrit commented on test/e2e/far_e2e_test.go:
```diff
@@ -403,13 +465,18 @@ func verifyStatusCondition(nodeName, conditionType string, conditionStatus *meta
 
 // checkRemediation verify whether the node was remediated
-func checkRemediation(nodeName string, nodeBootTimeBefore time.Time, pod *corev1.Pod, strategy v1alpha1.RemediationStrategyType) {
+func checkRemediation(nodeName string, nodeBootTimeBefore time.Time, pod *corev1.Pod, strategy v1alpha1.RemediationStrategyType, testAction string, ctx *e2eUtils.FenceCommandContext) {
```
is there any reason `ctx *e2eUtils.FenceCommandContext` is needed here ?
Can't we just pool this information from the FAR CR ?

kattz-kawa replied:
Originally, I used `ctx` to get the power status through the fencing command, because the FAR CR did not include reliable and direct power source information.
However, I found that `fence_aws --action=status` returns `Command Status: 2` when the power is OFF, which made it difficult to use.
So I changed the implementation of `GetPowerStatus` to retrieve the power status using the `machine-api` instead.
As a result, `ctx` is no longer needed, and I’ve removed it.

mshitrit commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +221,134 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+}
+
+// PowerOnNode executes the power-on command to turn on the node during the cleanup phase for the off-action test.
+func PowerOnNode(ctx *FenceCommandContext, targetNodeName string) (bool, error) {
```
Hmm I wonder if this is working on our AWS CI :thinking: , have you had a chance to verify  ?

mshitrit replied:
ATM e2e test [fails](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/medik8s_fence-agents-remediation/175/pull-ci-medik8s-fence-agents-remediation-main-4.19-openshift-e2e/1952745856759238656), that's something that will need fixing before we can merge.

kattz-kawa replied:
Do you think it would be reasonable to extend `verifyNodePoweredOff` timeout (`timeoutPowerOff`) as a fix?

I reviewed the must-gather logs and found that the test failure appears to be caused by a delay in the power-off operation of fence_aws, which is triggered by the far-controller. The command seems to have been retried several times before the node was actually powered off. Due to this delay, the `verifyNodePoweredOff` function hit its timeout, resulting in the test failure.

The simplest fix would be to increase the timeout value. However, I'm still unclear why the fence_aws command needed multiple retries in the CI environment.

Before proceeding with any changes, I’d appreciate your thoughts. Do you think extending the timeout is the right approach?

kattz-kawa commented on test/e2e/utils/command.go:
```diff
@@ -26,8 +26,23 @@ import (
 )
 
+// FenceCommandContext holds the information needed to run fence commands on the FAR controller pod
+type FenceCommandContext struct {
```
I introduced 'FenceCommandContext' struct to avoid passing multiple common/shared dependencies repeatedly across test helpers like `'wasNodePoweredOff'`, `'GetPowerStatus'`, and `'PowerOnNode'`.
This should help reduce repeated code, improve readability, and make the test code easier to maintain.

That said, maybe I'm missing a simpler way to achieve the same goal. If there's a cleaner pattern, like leveraging `context.Context` more effectively or pulling more from the FAR CR, I'd love to learn from that.

Let me know if you think this approach makes sense, or if you'd prefer a different direction.

mshitrit replied:
Sure, I'll try to explain myself better.

My point is that it seems like `FenceCommandContext` is a placeholder for the data which is fetched from `FenceAgentsRemediation`.
Why not just use `FenceAgentsRemediation` ?

In case there is still data that isn't part of  `FenceAgentsRemediation` you can store it on  `context.Context` .
[Here](https://github.com/medik8s/node-healthcheck-operator/blob/main/controllers/nodehealthcheck_controller.go#L161) is an example.

kattz-kawa replied:
Thanks for the clarification! That makes sense.  
I'll refactor accordingly and remove the `FenceCommandContext` struct.  
Appreciate the example, it really helped me understand the intended direction.

kattz-kawa replied:
I’ve updated the implementation. I removed the custom struct, as it is no longer necessary.

mshitrit commented:
/test 4.19-openshift-e2e

kattz-kawa commented:
> Sorry, but that is something you need to do yourself. Just rebase on your machine and force push the rebased code to your PR branch.

Got it, thanks for confirming.
I'll take care of the rebase and force-push the updated branch shortly.

mshitrit commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +223,106 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+
+// PowerOnNode executes the power-on command to turn on the node during the cleanup phase for the off-action test.
+func PowerOnNode(ctx context.Context, far *v1alpha1.FenceAgentsRemediation, targetNodeName string) (bool, error) {
+	clientset := ctx.Value(ClientSetContextKey).(*kubernetes.Clientset)
```
I think it's fine to pass `clientset` as a param to this method, it'll also keep consistency with the rest of the methods in this file.

kattz-kawa replied:
I was a bit too focused on using `ctx` and didn’t consider the simpler approach.
Thanks again for pointing it out, I've updated the code accordingly.

kattz-kawa commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +223,106 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+// PowerOnNode executes the power-on command to turn on the node during the cleanup phase for the off-action test.
+func PowerOnNode(ctx context.Context, far *v1alpha1.FenceAgentsRemediation, targetNodeName string) (bool, error) {
+	clientset := ctx.Value(ClientSetContextKey).(*kubernetes.Clientset)
+	controllerNS := ctx.Value(ControllerNSContextKey).(string)
```
I've updated the code.

mshitrit commented:
/test 4.19-openshift-e2e

mshitrit commented on test/e2e/far_e2e_test.go:
```diff
@@ -40,13 +39,15 @@ const (
 )
 
 var (
+	fenceAgentDefaultAction          string = "reboot"
```
Nit: this can be const

kattz-kawa replied:
I’ve moved it.

mshitrit commented on test/e2e/utils/command.go:
```diff
@@ -26,6 +26,9 @@ import (
 
 const (
 	containerTestName = "test-command"
+	// NOTE: These constants are duplicated from far_e2e_test.go for simplicity.
```
>I wasn’t sure if importing from the test file would be appropriate

I agree that production code shouldn't use test code variables, but in this case it's all test code so in theory I don't see any issue here.

However IIRC go will block you from importing variables form `_test`  since it woudn't make that distinction and built to prevent mix of test code variables into production code.

IMO the simplest way to solve this would be to remove the variables in `far_e2e_test.go` and instead import them from `command.go`

kattz-kawa replied:
I’ve removed the variables in `far_e2e_test.go` and defined in `command.go`

kattz-kawa commented on test/e2e/far_e2e_test.go:
```diff
@@ -376,6 +397,23 @@ func wasNodeRebooted(nodeName string, nodeBootTimeBefore time.Time) {
 }
 
+// wasNodePoweredOff checks if the node remains powered off
+func wasNodePoweredOff(nodeName string) {
```
Thanks for the suggestion.

Just to mention: I named it `wasNodePoweredOff` to keep consistent with the existing `wasNodeRebooted`, which follows a similar pattern and behavior. I totally understand that `verifyNodePoweredOff` might be clearer in isolation, though.

If you think renaming it would improve readability, I will update it, just let me know what you prefer.

kattz-kawa replied:
I've updated both.

clobrano commented on controllers/fenceagentsremediation_controller.go:
```diff
@@ -462,7 +469,7 @@ func (r *FenceAgentsRemediationReconciler) buildFenceAgentParams(ctx context.Con
 	for paramName, nodeMap := range far.Spec.NodeParameters {
 		if nodeVal, isFound := nodeMap[v1alpha1.NodeName(nodeName)]; isFound {
 			// Verify action must be reboot
```
This should be "reboot or off"

kattz-kawa replied:
Thanks, I've updated it to 'reboot or off'.

clobrano commented on controllers/fenceagentsremediation_controller.go:
```diff
@@ -480,7 +487,7 @@ func (r *FenceAgentsRemediationReconciler) buildFenceAgentParams(ctx context.Con
 	for secretKey, secretVal := range secretParams {
 		secretParam := v1alpha1.ParameterName(secretKey)
 		// Verify action must be reboot
```
This should be "reboot or off"

kattz-kawa replied:
Thanks, I've updated it to 'reboot or off'.

clobrano commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +216,101 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+		}
+
+	case fenceAgentIPMI:
+		var username, password string
```
Please, keep the "TODO" comments we have around this logic in main

kattz-kawa replied:
Done, TODOs are kept.

clobrano commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +216,101 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+	switch agent {
+	case fenceAgentAWS:
+		var accessKey, secretKey string
+		accessKey, secretKey, err = GetSecretData(c, "aws-cloud-fencing-credentials-secret", "openshift-operators", "aws_access_key_id", "aws_secret_access_key")
```
Please, use constants for this values. Same below for BMC.

clobrano reviewed (changes requested):
I left some comments

mshitrit commented on test/e2e/far_e2e_test.go:
```diff
@@ -107,17 +109,23 @@ var _ = Describe("FAR E2e", func() {
 			DeferCleanup(deleteFAR, far)
+			// The node needs to be powered on after 'off' remediation
+			if testAction == "off" {
+				DeferCleanup(e2eUtils.PowerOnNode, clientSet, far, selectedNode.Name, log)
```
Nit: event though this should be triggered fine due to Defer stacking (i.e `waitForNodeHealthyCondition` will be triggered after `PowerOnNode`) IMO this is still quite confusing.
Since this is the only place `PowerOnNode` I think it'll be simpler just to add the `waitForNodeHealthyCondition` inside the `PowerOnNode` and have only one `DeferCleanup` triggered.

kattz-kawa replied:
Thanks, I introduced `powerOnNodeAndWaitUntilReady` to improve the readability of the execution flow between `PowerOnNode` and `waitForNodeHealthyCondition`.

mshitrit commented on test/e2e/utils/command.go:
```diff
@@ -205,3 +216,101 @@ func GetPod(nodeName, containerName string) *corev1.Pod {
+	log.Info("Output of power-on command", "stdout", string(bytes))
+
+	if err != nil {
+		log.Error(err, "Power-on Failed")
```
Nits:
- Since not params are used in the error text I think `errors.New` is better than `fmt.Errorf`
- error is expected to start with lower case

kattz-kawa commented:
> don't forget to update the README please :)

@slintes
I'm working on updating the README. Rather than modifying all existing mentions of "reboot", I'm planning to add a **separate section specifically for the newly supported `off` action**, explaining its purpose and usage. I think this approach will keep the documentation clearer and easier to maintain. Does that sound good to you? 

In parallel, I'm also addressing your comment by adding the timeout and removing unnecessary variables. 
Let me know if you have any preferences or suggestions. Thank you.

kattz-kawa commented on controllers/fenceagentsremediation_controller.go:
```diff
@@ -52,11 +52,14 @@ const (
+	SuccessFAResponse           = "Success: Rebooted"
+	parameterActionName         = "--" + actionName
+	actionName                  = "action"
+	parameterDefaultActionValue = "reboot"
```
I've updated that code.

mshitrit commented:
/test 4.19-openshift-e2e

clobrano reviewed (approved):
I think I am the last one

/lgtm

razo7 commented:
Very good PR @kattz-kawa, and I have allowed all the testing.
I have some thoughts on small NITs, but that should not block the progress of this PR.

Please remove [WIP] in the PR's title for merging the PR :)

razo7 commented on README.md:
```diff
@@ -10,19 +10,32 @@ FAR uses a fence agent to fence a Kubernetes node. Generally, fencing is the pro
+
+If you choose the `off` action, FAR powers off the unhealthy node using the fence agent. Unlike `reboot`, the node does not automatically return to service. It remains powered off until manually recovered by an administrator. The taint and workload eviction behavior remain the same as in the `reboot` case, ensuring that the node is isolated and workloads are safely rescheduled elsewhere.
 
 FAR includes the `FenceAgentsRemediationTemplate` (or `fartemplate`) Custom Resource Definition ([CRD](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)) for how to create a FenceAgentsRemediation CR. The template has the same fields as far CR (e.g., agent name) and it is used for automatically creating remediation  ...
```
NIT: The following line could be interpreted under the new `### Power-off-based remediation` subsection.

It might be better to have a new subsection for  `fartemplate` so this line won't be considered as only associated with power-off remediation.

kattz-kawa commented:
@slintes 
> Patching the machine for forcing a reconcile... hm, not sure, might be worth a try.

I’ve just committed an implementation that patches the Machine to force a reconcile. 
Could you please take a look and review it?

In my environment, all tests are passing, and the remediation finish time has improved. Here are results:

```
2025-08-20T08:28:00.672750166+09:00     INFO    Remediation was finished        {"remediation time": "3m0.79322408s"}
2025-08-20T08:28:00.672815001+09:00     INFO    Remediation was finished        {"remediation time": "1m59.614899288s"}
2025-08-20T08:28:00.672827852+09:00     INFO    Remediation was finished        {"remediation time": "3m10.676062952s"}
2025-08-20T08:28:00.672849274+09:00     INFO    Remediation was finished        {"remediation time": "1m51.687023791s"}
2025-08-20T08:28:00.67285381+09:00      INFO    Remediation was finished        {"remediation time": "1m36.581419788s"}
2025-08-20T08:28:00.672858347+09:00     INFO    Remediation was finished        {"remediation time": "1m2.099814578s"}
2025-08-20T08:28:00.672862562+09:00     INFO    Remediation was finished        {"remediation time": "1m38.030299522s"}
2025-08-20T08:28:00.672867499+09:00     INFO    Remediation was finished        {"remediation time": "58.323772848s"}
(snip)
[AfterSuite] PASSED [0.000 seconds]
------------------------------

Ran 8 of 8 Specs in 1577.180 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

razo7 commented on test/e2e/far_e2e_test.go:
```diff
@@ -403,13 +448,22 @@ func verifyStatusCondition(nodeName, conditionType string, conditionStatus *meta
+		// After detecting the node power-off, the Pod may not be immediately deleted.
+		// Calling checkPodDeleted too early can lead to fail test.
+		// Therefore, we wait until the Pod is removed before verifying its deletion status.
+		waitForPodDeletion(pod)
```
>This looks fishy to me. Why would there be a difference to the reboot case? I don't remember that the test ever failed at checkPodDeleted(). @razo7 do you?

I don't remember as well

>However, I'm wondering why checkPodDeleted uses Consistently, and not Eventually? 🤔 @razo7 do you remember why?

No, and it makes sense to use Eventually

kattz-kawa commented on test/e2e/utils/cluster.go:
```diff
@@ -120,3 +132,95 @@ func GetBMHNodeInfoList(machineClient *machineclient.Clientset) (map[v1alpha1.No
+// GetPowerStatus retrieves the node's power status by querying the corresponding Machine object.
+// To ensure the status is current, this function first triggers a reconciliation of the Machine resource.
+// This helps speed up synchronization with the machine-api controller before reading the providerStatus.
+
```
I've updated.

kattz-kawa commented on test/e2e/utils/cluster.go:
```diff
@@ -120,3 +132,95 @@ func GetBMHNodeInfoList(machineClient *machineclient.Clientset) (map[v1alpha1.No
+		return "", fmt.Errorf("failed to get machine name %s: %w", targetNodename, err)
+	}
+	// Trigger reconcile for status
+	TriggerReconcileMachine(machineClient, machinename)
```
I've updated.
